stl2ps.1
.\" -*- nroff -*-
.\" stl2ps.1
.\" By: R.F. Smith <rsmith@xs4all.nl>
.\" $Date$
.Dd
.Os 
.Dt STL2PS 1 CON
.Sh NAME
.Nm stl2ps
.Nd Projects an STL file on the (x,y) plane and renders a PostScript figure of it.
.Sh SYNOPSIS
.Nm 
infile
.Op Ar outfile
.Op x Ar degrees | y Ar degrees | z Ar degrees ...
.Sh DESCRIPTION
This manual page documents  
.Ql $Revision$
of the 
.Nm
program. This program's purpose is to render a globally illuminated grayscale
projection of an STL file as PostScript output.

First, the STL file is parsed. Both binary and text STL file can be read, but
any color information in binary STL files is discarded since this is not
standardized. While STL files are supposed to contain closed solids, by design
no effort is made to try and repair gaps in the STL file. There simply isn't
enough geometric context to do this in a foolproof way.

Next, all the triangles in the file are rotated around the axis of the global
coordinate system, as specified by the sequence of rotation options given on
the command line.

Finally, the visible triangles (visibility is determined by a non-negative
z-component of the unit normal vector of the triangle) are sorted by depth,
projected on the (x,y) plane. So in the PostScript picture you are looking
along the z-axis to the origin. The picture is scaled to fit in a maximum size
of 100x100 mm and the triangles are drawn from back to front using the
.Ql gfill
operator. The grayscale tint of the triangle is also determined by the
orientation of the triangle; running from almost black for triangles turned 90
degrees to the line of sight to near white for triangles parallel to the x,y
plane.

If no output filename is given, the path and the
.Ql .stl 
extension are stripped from the input filename. The extension is replaced by the
.Ql .ps
extension. So if no output filename is specified, the output file is written
in the current working directory. The PostScript data is written to the output
filename.
.Sh EXIT STATUS
.Ex -std
.Sh DIAGNOSTICS
The following non-fatal warnings can be issued when the program cannot make
sense of command line arguments:
.Bl -tag -width "foo"
.It Unknown argument '...' ignored.
One of the command line transform arguments did not start with one of the
letters 'x', 'y', 'z', 'X', 'Y' or 'Z' and will be ignored. 
.It Argument '...' is not a number, ignored.
A non-numeric value was encountered among the command line arguments while a
number of degrees is expected as part of a rotation command. It will be ignored.
.El
.Pp
There are some errors which the program cannot ignore, and which will
terminate the program.
.Bl -tag -width "foo"
.It The file '...' cannot be read or parsed. Exiting.
If a given input file cannot be read, or if the file is not recognized as an
STL file, this error is produced, and the program is terminated with exit code
1.
.It Cannot write output file '...'.
An output file name was given on the command line, but the file in question
could not be opened for writing or written to. The program is terminated with
exit code 2.
.El
.Sh COMPATIBILITY
The 
.Nm
program requires the Python interpreter. It was written for version 2.7, but
should be able to work with 3.x after fixing with
.Xr 2to3 .
.Sh SEE ALSO
.Xr python 1 ,
.Xr stl2pov 1 ,
.Xr 2to3.
.Rs
.%B Python Programming Language - Official Website
.%U http://www.python.org/
.Re
.Sh HISTORY
The origin of this software was found in the desire of the author to render 3D
CAD models using the POV-ray raytracer. The first effort produced the
.Xr stl2pov 1
program, written in C. 
.Pp
Later the author converted that program to Python as a learning
experience. This resulted in a more re-usable version of the software to parse
STL files as a Python module. Adding this front-end for outputting a
PostScript rendering then became much easier.
.Sh AUTHOR
This manual and the 
.Nm
software were written by 
.An Roland Smith Aq rsmith@xs4all.nl .
.Pp
The latest version of this program is available at:
.Lk http://rsmith.home.xs4all.nl/software/ 
.Sh LICENSE
To the extent possible under law, Roland Smith has waived all copyright and
related or neighboring rights to this manual. This work is published from the
Netherlands. See 
.Lk http://creativecommons.org/publicdomain/zero/1.0/
.Pp
The
.Nm
program itself is released under the two-clause BSD license given below;
.Pp
Copyright \(co 2012 R.F. Smith <rsmith@xs4all.nl>. All rights reserved.
.Pp
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
.Bl -tag -width "foo" -compact
.It 1.
Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
.It 2.
Redistributions in binary form must reproduce the above copyright notice, this
list of conditions and the following disclaimer in the documentation and/or
other materials provided with the distribution.
.El
.Pp
THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS
OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" EOF


stl2ps.py
#! /usr/bin/env python
# -*- python coding: utf-8 -*-
# Copyright © 2012 R.F. Smith <rsmith@xs4all.nl>. All rights reserved.
# $Date$
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.

'''Program for converting a view of an STL file into PostScript output.'''

import os
import sys
import time

import stl
import xform

name = ('stl2ps [ver. ' + '$Revision$'[11:-2] + 
       '] ('+'$Date$'[7:-2]+')')

def usage():
    print name
    print "Usage: stl2ps infile [outfile] [transform [transform ...]]"
    print "where [transform] is [x number|y number|z number]"

## This is the main program ##
# Process the command-line arguments
validargs = ['x', 'y', 'z', 'X', 'Y', 'Z']
if len(sys.argv) == 1:
    usage()
    sys.exit(0)
infile = sys.argv[1]
if len(sys.argv) < 3 or sys.argv[2] in validargs:
    outfile = None
    del sys.argv[:2]
else:
    outfile = sys.argv[2]
    del sys.argv[:3]
tr = xform.Xform()
while len(sys.argv) > 1:
    if not sys.argv[0] in validargs:
        print "Unknown argument '{}' ignored.".format(sys.argv[0])
        del sys.argv[0]
        continue
    try:
        ang = float(sys.argv[1])
        if sys.argv[0] in ['x','X']:
            tr.rotx(ang)
        elif sys.argv[0] in ['y','Y']:
            tr.roty(ang)
        else:
            tr.rotz(ang)
        del sys.argv[:2]
    except:
        print "Argument '{}' is not a number, ignored.".format(sys.argv[1])
        continue
# Open the file
try:
    stlobj = stl.Surface(infile)
except:
    print "The file '{}' cannot be read or parsed. Exiting.".format(infile)
    sys.exit(1)
# Process the file
for result in stlobj.processfacets:
    print result
# Remove spaces from name
stlobj.name = stlobj.name.strip()
# Apply transformations
if tr.unity == False:
    stlobj.xform(tr)
# Calculate viewport and transformation
xmin, xmax, ymin, ymax, zmin, zmax = stlobj.extents()
pr = xform.Zpar(xmin, xmax, ymin, ymax)
# Prepare output string.
outs = "%!PS-Adobe-1.0\n"
outs += "%%BoundingBox: 0 0 {:.0f} {:.0f}\n".format(pr.w, pr.h)
outs += "% Generated by {} on {}.\n".format(name, time.asctime())
outs += stlobj.stats('% ')+'\n'
s = "% The scale factor used is: {} PostScript points/STL-unit\n"
outs += s.format(pr.s)
s = "% This becomes a picture of {:.0f}×{:.0f} PostScript points;"\
    " {:.0f}×{:.0f} mm.\n"
outs += s.format(pr.w, pr.h, pr.w/72*25.4, pr.h/72*25.4)
# Calculate the visible facets
vizfacets = [f for f in stlobj.facets if pr.visible(f.n)]
outs += "% {} of {} facets are visible.\n".format(len(vizfacets), len(stlobj))
# Next, depth-sort the facets using the smallest distance to the viewer
# of the three vertices.
vizfacets.sort(None, lambda f: max([f.v[0].z, f.v[1].z, f.v[2].z]))
# PostScript settings and macros.
outs += ".5 setlinewidth\n"
outs += "/g {setgray} def\n"
outs += "/f {moveto} def\n"
outs += "/s {lineto} def\n"
outs += "/t {lineto closepath gsave fill grestore stroke} def\n"
# Project and illuminate the facets
pf = (stl.ProjectedFacet(f, pr) for f in vizfacets)
# Draw the triangles
for f in pf:
    s = "{:4.2f} g {:.3f} {:.3f} f {:.3f} {:.3f} s {:.3f} {:.3f} t\n"
    outs += s.format(f.gray, f.x1, f.y1, f.x2, f.y2, f.x3, f.y3)
# Showpage must be the last line in the PostScript output.
outs += "showpage\n"
# Send output.
if outfile == None:
    # derive output filename from input filename.
    outbase = os.path.basename(infile)
    if outbase.endswith((".stl", ".STL")):
        outbase = outbase[:-4]
    outfile = outbase+".ps"
try:
    outf = open(outfile, "w+")
    outf.write(outs)
    outf.close()
except:
    print "Cannot write output file '{}'".format(sys.argv[2])
    sys.exit(2)
